### PR TITLE
New Module: win_path

### PR DESCRIPTION
**Pull request by nitzmahone**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_path

##### SUMMARY
Add win_path module for idempotent management of Windows path env vars

### CODE REVIEW

**Comment by nitzmahone** [comment]
cc @jhawkesworth @dagwieers

**Comment by dagwieers** [comment]
My first impression by the name 'win_path' is that it would manage paths, not PATH envs.
So rather than paths it manages environment variables, so I wonder, are there other environment variables that could be managed the same way ? And shouldn't it become win_env instead, but we already have a win_environment. Oh the confusion :-)

Maybe win_pathenv or win_env_path or simply ignore my ignorance :-D

**Comment by nitzmahone** [comment]
@dagwieers It defaults to `PATH`, but will manage any var you want (see the `name` arg) that corresponds to the Windows path behavior (eg, Java CLASSPATH, etc) with semicolon-separated paths.

**Comment by nitzmahone** [comment]
I tried to think of a sane way to jam this behavior into the existing `win_environment`, but I think it'd be another thing we'd chalk up to "a module trying to do too much" with polymorphic UI and such...

**Comment by dagwieers** [comment]
Yeah, I don't think it belongs in win_environment either. And my ramblings was my first impression. I did read the module docs eventually. But I am sure others will have the same confusion when skimming the list of modules. (So feel free to ignore if this is a non-issue.)

**Comment by nitzmahone** [comment]
I could get behind `win_env_path`- let's see if anyone else feels like weighing in... @trondhindenes @bcoca @jborean93

**Comment by dagwieers** [comment]
The win_environment module uses `level:` rather than `scope:`, however I think `scope:` is a much better term. So I would propose to also change the default to `scope:` in win_environment when this is accepted.

I find the term `machine` also strange, I never use that before. (except for physical things) I'd rather go for `system`. But I noticed that win_environment is doing the same thing so maybe this is Windows/Powershell-terminology or maybe I am weird :-)

Now that I think of it VM means virtual machine, so... :-)

**Comment by nitzmahone** [comment]
Yeah, `Machine` is the terminology used by the underlying API: https://msdn.microsoft.com/en-us/library/system.environmentvariabletarget(v=vs.110).aspx, so I stuck with that.

**Comment by trondhindenes** [comment]
I agree with "scope" being a good param name. Nice addition to the Windows support in Ansible, I'm all for it! Two enthusiastic thumbs up!

File: lib/ansible/modules/windows/win_path.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'core',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: win_path
+version_added: "2.3"
+short_description: Manage Windows path environment variables.
```
Comment (gundalow) [inline lib/ansible/modules/windows/win_path.py]: `short_description` without full stop. All other `description:` with full stop

File: lib/ansible/modules/windows/win_path.ps1 [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+# calculate the new path value from the existing elements
+$path_value = [String]::Join(";", $existing_elements.ToArray())
+$result.path_value = $path_value
+
+If ($result.changed -and -not $check_mode) {
+    Set-RawPathVar $path_value $scope | Out-Null
+}
+
+Exit-Json $result
```
Comment (gundalow) [inline lib/ansible/modules/windows/win_path.ps1:158]: Missing newline.

**Comment by jhawkesworth** [comment]
I sort of wish I hadn't grabbed the name win_environment now.
I think I'd prefer the name win_env_var because there's nothing really special about PATH other than what it gets used for. Its worth having discoverable names I think.

I probably used 'level' because it was the microsoft term as well, I don't recall now, but I don't feel strongly about that, 'scope' makes just as much sense to me.  Could alias param name 'level' to 'scope' in win_environment although I can't really think of a reason to keep win_environment long term, so perhaps just deprecate it in favour of this.

**Comment by nitzmahone** [comment]
@jhawkesworth Nah, I think `win_environment` is still important for non-PATH-y kinds of things- this is just more flexible for managing individual elements of PATH-shaped vars. I think they're complementary...

**Review by gundalow — DISMISSED** [review]
needs_revision

Might be good to do
win_path add
Check 
win_path add 2nd time & assert no change # < - New test

Similarly for state: absent

**Comment by gundalow** [comment]
Regarding `win_environment` vs `win_path` I wonder if it would be worth adding a `note:` in each to refer to the the module and say when one should be used vs other?

**Comment by jhawkesworth** [comment]
Ok I see the distinction now, I've fine with calling it win_path in that case and like @gundalow 's suggestion of adding a note to both this and win_environment about intended usage.  

To me its all just setting key=value although when I looked back at playbooks here I only actually use win_environment to set things that look like paths.  

I've said elsewhere that I think setting half the path in your playbook and leaving the rest up to what's already on the target machines is a bad idea (its recipe for snowflake systems and hard to discover configuration), but ansible shouldn't be prescriptive about such things - you can't always automate everything after all.

**Comment by nitzmahone** [comment]
@jhawkesworth exactly- otherwise we wouldn't have things like `lineinfile` (which I agree is often a devops-smell); this one in particular though can also be good for helping prevent people from shooting themselves in the foot by, say, removing Powershell from the path. :)

**Pull request merged**